### ISSUE TITLE
Spring-Boot-Starter 2.5.0 and junit

### ISSUE THREAD

**Issue by DoctorHowser** [togglz/togglz#261]
Hey Folks,

TL:DR Anyone have a guide or reference for dealing with a FeatureManager thats @Autowired for running junit tests?

We have a standard spring-boot-starter use of Togglz, where we autowire the FeatureManager and use it where we need.

  @Autowired
    private FeatureManager manager

This works fine. Where I cant seem to find any docs is the usage in setting up unit tests (junit). Before, we could do something like:

  @Rule
  public TogglzRule togglzRule = TogglzRule.allEnabled(MyFeatures.class);

but this doesn't instantiate the FeatureManager, so when the test runs, the manager is null. As I would expect, since @autowired isn't run in tests. 

We tried manually instantiating the FeatureManager for the test, which works, except it seems that we can only set something to override the default isActive() method, so its not able to be changed--or put another way, TogglzRule does nothing.

Any thoughts?

**Comment by dasniko** [comment]
This is not a Togglz related thing, but generally a wrong usage of field injection. Field injection is deprecated and evil!

Use constructor injection instead, an testing will get a lot easier.
You then can, for example, mock your `FeatureManager` with Mockito and define the behaviour of this mock for your tests. Then instantiating your class, just put the mock in the constructor call - and you're good to go. No need of workarounds that evil field injection will work.

See also here: http://olivergierke.de/2013/11/why-field-injection-is-evil/

**Comment by DoctorHowser** [comment]
I get that, but also the guide for togglz spring boot specifically has you @Autowired the manager. 
https://www.togglz.org/documentation/spring-boot-starter.html

I kinda think the Togglz docs for spring boot should then have some sort of guide for testing with the @Autowired annotation as dictated, or the testing section should make note of integrating with spring boot?

If @dasniko's suggestion is to be followed, then the spring-boot docs need to be changed to use constructor injection, but I'm not qualified to make that choice.

**Comment by dasniko** [comment]
@chkal WDYT?
I could update the Spring Boot docs, regarding _not_ to use `@Autowired`.

The Testing docs are not touched/modified since 5 yrs, says commit history.
Do you plan anything to update this anyway? Or should we include some Spring Testing support in there?

If so, I'd recommend to create two separate issues. One for updating the Spring Boot docs, another for Testing.

**Comment by chkal** [comment]
Hey all. Sorry for the delayed response.

First let me describe why `@Autowired` doesn't work very well in your case. Basically the Togglz testing support is based on the idea of providing a `FeatureProvider` that returns a dummy `FeatureManager` which can be configured with `TogglzRule`. However, `FeatureProvider` is only used to lookup the `FeatureManager` if you use `FeatureContext.getFeatureManager()`, which is the case if you use `MyFeature.FEATURE1.isActive()` for checking features. If you inject `FeatureManager` using Spring's `@Autowired`, you will basically bypass `FeatureProvider` and directly get the standard `FeatureManager` you are using in your application and which is managed by Spring.

I know that everyone says dependency injection is a good thing and I totally agree with this. However, for Togglz this **may** be a different thing. I spent quite some time to implement the dynamic `FeatureManager` lookup that is happening if you call `MyFeature.FEATURE1.isActive()` and it works great even in different environments (Spring, CDI, Guice, plain Java, etc) and in complex classloader scenarios. Especially it also works in places where `@Autowired` is not available like static utility methods, value objects and so on.

Therefore I think we should just update the docs of the Spring Boot module accordingly. The testing docs are very old and maybe not up to date any more. It should definitevely be updated. But perhaps we should wait with this until we get #258 in, which I didn't find time for to review yet. Help welcome. ;-)

**Comment by DoctorHowser** [comment]
Thanks @dasniko , @chkal !
